File: python/sort/sort.py


m = [2, 3, 3, 7, 3, 9, 9, 12, 6]
n = [1, 3, 4, 7, 10, 16, 12, 9, 17, 19]
a = m + n
n = len(a)

def bubble_sort():

    for i in range(0, n):
        for j in range(0, n-i-1):
            if a[j] > a[j+1]:
                tmp = a[j+1]
                a[j+1] = a[j]
                a[j] = tmp

    print('new array:')
    print(a)

# a = [1, 4, 7, 3, 11, 9, 2, 6, 4, 7]
def insert_sort():
    for i in range(1, n):
        tmp = a[i]
        j = i - 1
        
        while j > -1:
            if tmp < a[j]:
                a[j+1] = a[j]
                j -= 1
            else:
                break

        a[j+1] = tmp

    print('new array:')
    print(a)

# a = [1, 4, 7, 3, 11, 9, 2, 6, 4, 7]
def select_sort():
    for i in range(0, n):
        min = a[i]
        t = i

        for j in range(i+1, n):
            if min > a[j]:
                min = a[j]
                t = j

        if t != i:
            a[t] = a[i]
            a[i] = min

    print('new array:')
    print(a)


def main():
    # bubble_sort()
    insert_sort()
    # select_sort()

if __name__ == '__main__':
    main()

File: c/sort/sort.c
#include <stdio.h>
#include <stdlib.h>

#define N 10

int a[N] = {7, 6, 3, 8, 15, 6, 7, 3, 12, 10};

void bubble_sort() {

    for(int i = 0; i < N; i++) {
        for (int j = 0; j < (N-i-1); j++) {
            if (a[j] > a[j+1]) {
                int tmp = a[j];
                a[j] = a[j+1];
                a[j+1] = tmp;
            }
        }
    }

    printf("new array: \n");
    printf("    ");
    for (int i = 0; i < N; i++)
        printf("%d ", a[i]);
    printf("\n");
}

// int a[N] = {7, 6, 3, 8, 15, 6, 7, 3, 12, 10};
void insert_sort() {

    for(int i = 1; i < N; i++) {
        int tmp = a[i];
        printf("tmp = %d\n", tmp);
        int j = i - 1;
        for (j = i-1; j >= 0; j--) {
            printf("a[i] = %d, a[j]=%d\n", a[i], a[j]);
            if (tmp < a[j]) {
                a[j+1] = a[j];
            } else {
                break;
            }
        }

        a[j + 1] = tmp;
    }

    printf("new array: \n");
    printf("    ");
    for (int i = 0; i < N; i++)
        printf("%d ", a[i]);
    printf("\n");
}

// int a[N] = {1, 6, 3, 8, 15, 6, 7, 3, 12, 10};
void select_sort() {

    for(int i = 0; i < N; i++) {
        int min = a[i];
        int t = i;
        for (int j = (i+1); j < N; j++) {
            if (a[j] < min) {
                min = a[j];
                t = j;
            }
        }

        if (t != i) {
            a[t] = a[i];
            a[i] = min;
        }
    }

    printf("new array: \n");
    printf("    ");
    for (int i = 0; i < N; i++)
        printf("%d ", a[i]);
    printf("\n");
}

int main() {

    // bubble_sort();
    insert_sort();

    return 0;
}